src/kOS.Safe/Utilities/AssemblyWalkAttribute.cs

```diff
@@ -0,0 +1,302 @@
+
+        public AssemblyWalkAttribute()
+        {
+            LoadedCount = 0;
+        }
+
+        public override string ToString()
+        {
+            string format = "AssemblyWalkAttribute({0})";
+            if (AttributeType != null)
+                return string.Format(format, AttributeType.FullName);
+            if (InterfaceType != null)
+                return string.Format(format, InterfaceType.FullName);
+            if (InherritedType != null)
+                return string.Format(format, InherritedType.FullName);
+            if (!string.IsNullOrEmpty(StaticWalkMethod))
+                return string.Format(format, StaticWalkMethod);
+            return string.Format(format, "<None>");
+        }
+
+        public static void Walk()
+        {
+            SafeHouse.Logger.Log("AssemblyWalkAttribute begin walking");
+            LoadWalkAttributes();
+            WalkAllAssemblies();
+        }
+
+        public static void LoadWalkAttributes()
+        {
+            AllWalkAttributes.Clear();
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            foreach (Assembly assembly in assemblies)
+            {
+                // Don't check the assembly if it's in the GAC.
+                // This is mostly to save time, since there are a number of assemblies
+                // in the GAC, and it should not be possible for them to at all implement
+                // this attribute.
+                if (!assembly.GlobalAssemblyCache)
+                {
+                    SafeHouse.Logger.SuperVerbose("Loading assembly: " + assembly.FullName);
+                    try
+                    {
+                        LoadWalkAttribute(assembly);
+                    }
+                    catch
+                    {
+                        SafeHouse.Logger.LogWarning("Error while loading assembly: " + assembly.FullName + ", skipping assembly.");
+                    }
+                }
+            }
+            SafeHouse.Logger.SuperVerbose("Loaded " + AllWalkAttributes.Count + " attributes.");
+        }
+
+        public static bool CheckMethodParameters(Type baseType, string methodName, params Type[] parameterTypes)
+        {
+            var managerRegisterMethod = baseType.GetMethod(methodName, BindingFlags.Static | BindingFlags.Public);
+            if (managerRegisterMethod != null)
+            {
+                // Check the register method to make sure it supports both parameters.
+                var parameters = managerRegisterMethod.GetParameters();
+                if (parameters.Length != parameterTypes.Length)
+                {
+                    return false;
+                }
+                // We don't need to check the parameter types if there are no parameters.
+                if (parameters.Length == 0)
+                {
+                    return true;
+                }
```

Comment:
As a rule of thumb, I don't like exception cases that say "avoid the loop when collection is empty", if the loop is already written as a "do the loop once per element of the collection".  "avoid it when it's empty" is sort of already built-in in that case.